fix: insomnia open dialog for proto directory can't select directories

koppathachan opened the pull request:
refer: [Issue](https://github.com/Kong/insomnia/issues/3347)
This change seems to fix above issue, it might not be right.

Closes #3347, closes #3420, closes INS-741

koppathachan committed:
fix: insomnia open dialog for proto directory can't select directories

develohpanda commented:
Thanks for checking this on several OS's and raising a PR! Pinging @dimitropoulos to explore whether this is the best solution, as it seems to be specific to Linux.

Excellent candidate for a smoke test as well, in future.

dimitropoulos committed:
uses a named export for selectFileOrFolder

(also, removes original js file from rebase)

dimitropoulos committed:
clears error by leveraging exhaustiveness check

dimitropoulos committed:
fixes bug: the `name` field is actually for a file filter

see the referenced pull request.

As for the `extensions: ['*']`, there's no reason I can see to include a filter and then tell the filter to then accept everything.

dimitropoulos commented:
That ended up being quite the adventure! :)

It's working now on.  @develohpanda passing the puck back to you to try re-verify on Mac after my Linux fix.  The `name` filed looks to have been misinterpreted before as being some kind of title for the dialog.  When, in fact, it was being used to allow or disallow certain files or folders from being selected.

For example if I:
```diff
- options.filters = [{
   extensions,
   
}];
+    options.filters = [{ name: 'ziltoid', extensions }];
```

Then I get a dialog like this:


Note the bottom right.  It says `ziltoid`.  It's looking for a file with that name.  I made one such file, and sure enough, it's selectable.  Now take a look back at the original issue and you'll see, likewise, that `Files` is what's filtered https://github.com/Kong/insomnia/issues/3347.

develohpanda committed:
update selectFileOrFolder mocks

develohpanda reviewed (approved):
This PR appears to be working on macOS as well 👍🏽 

It appears the following usage might also be broken on Linux for the same reason?

https://github.com/Kong/insomnia/blob/f2609fcaa9beaeacb73d477517a5d01956461997/packages/insomnia-app/app/ui/redux/modules/global.tsx#L369-L385

I believe you can reach this code path from Preferences > Data  > Import Data > From File.

dimitropoulos committed:
use switch (for exhaustiveness checking) and type selectedFormat

dimitropoulos committed:
removes unnecessary filters from _save_ dialog

from the docs:
> The filters specifies an array of file types that can be displayed

As suspected, this is not needed.  A user is free to save it wherever they want.

dimitropoulos committed:
adds extension to saved file

not sure why this was missing before, but it appears to have been a bug

dimitropoulos committed:
formatting updates

best to "ignore whitespace" for this commit.  I did this with the hope of using the `ThunkAction` type from `redux-thunk`, but once I got them all looking good and started adding the type I quickly learned there's quite a bit more work to do in this area before we can have such a thing.  I therefore opted to just call it a day at that and take the (no-op) formatting changes and typings.

dimitropoulos committed:
removes remaining name filters from save dialogs

same reason as the 2nd to prior commit - they cause the bug
